homeassistant/components/zha/entity.py

```diff
@@ -126,14 +125,9 @@ def async_set_state(self, state):
             None, "{}_{}".format(self.zha_device.available_signal, 'entity'),
             self.async_set_available,
             signal_override=True)
-        await self.async_accept_signal(
-            None, "{}_{}".format(SIGNAL_REMOVE, str(self.zha_device.ieee)),
-            self.async_remove,
-            signal_override=True
-        )
         self._zha_device.gateway.register_entity_reference(
-            self._zha_device.ieee, self.entity_id, self._zha_device,
-            self.cluster_channels, self.device_info)
+            self._zha_device.ieee, self, self.device_info
```

Comment:
Instead of storing the entity, we could store a `Future` that represents when the entity has been removed. The future is done when the entity has been removed. We can then wait for this future and any other needed futures to be done.

Reply:
Doesn’t that seem a bit over engineered to handle removes correctly? Keep in mind, the case we’re solving for is device removal. Each device has n many entities and we need to wait until all are removed before we can clean up the device and entity registries.

Reply:
I’m not totally following the futures thing... if you can provide an example I’m more than willing to give it a shot.